### PR TITLE
Pruning callback for tf.keras

### PR DESCRIPTION
**Pull request by sfujiwara**
I implemented `TFKerasPruningCallback` for #490.

* `tfkeras.py`
  * Implementation of `TFKerasPruningCallback`
* `tfkeras_integration.py`
  * An example of `TFKerasPruningCallback`
* `test_tfkeras.py`
  * Unit tests for `TFKerasPruningCallback`

The implementation is similar to [`KerasPruningCallback`](https://optuna.readthedocs.io/en/latest/reference/integration.html#optuna.integration.KerasPruningCallback).
It report the score to the trial every epoch.

## Maybe future works but out of scope for this PR

* This callback is intended to be compatible for TensorFlow v1 and v2, but tested only with v1
  * We need large modification to the config of CircleCI
* It's better to report score every n batch
  * Currently report only every epoch

### CODE REVIEW

File: examples/pruning/tfkeras_integration.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    ds = ds.repeat().batch(BATCHSIZE)
+    ds = ds.prefetch(tf.data.experimental.AUTOTUNE)
+
+    return ds
+
+
+def create_model(trial):
+    # type: (optuna.trial) -> tf.keras.Model
+
+    # Hyper parameters to be tuned by Optuna.
```
Comment (toshihikoyanase) [inline examples/pruning/tfkeras_integration.py]: Please replace `Hyper parameters` with `Hyperparameters` to unify the terms used in this repository.

File: examples/pruning/tfkeras_integration.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+BATCHSIZE = 128
+CLASSES = 10
+EPOCHS = 20
+N_TRAIN_EXAMPLES = 60000
+STEPS_PER_EPOCH = int(N_TRAIN_EXAMPLES / BATCHSIZE / 10)
+VALIDATION_STEPS = 30
+
+
+def train_dataset():
+    # type: () -> tf.data.Dataset
```
Comment (toshihikoyanase) [inline examples/pruning/tfkeras_integration.py]: Please remove type hints in this file. 
Optuna's examples do not have type hints for novice users to read the code easily.

File: optuna/integration/tfkeras.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+        Add a pruning callback which observes validation losses.
+
+        .. code::
+
+            model.fit(x, y, callbacks=[TFKerasPruningCallback(trial, 'val_loss')])
+
+    Args:
+        trial:
+            A :class: `~optuna.trial.Trial` corresponding to the current evaluation of the
```
Comment (toshihikoyanase) [inline optuna/integration/tfkeras.py]: Please remove the whitespace between `:class:` and `~optuna.trial.Trial` to enable the link to `Trial` properly.

File: examples/pruning/tfkeras_integration.py (hunk 1 above)
Comment (sfujiwara) [reply examples/pruning/tfkeras_integration.py]: I fixed it

File: examples/pruning/tfkeras_integration.py (hunk 2 above)
Comment (sfujiwara) [reply examples/pruning/tfkeras_integration.py]: I fixed it

File: optuna/integration/tfkeras.py (hunk 3 above)
Comment (sfujiwara) [reply optuna/integration/tfkeras.py]: I fixed it

**Comment by sfujiwara** [comment]
@toshihikoyanase Thanks! I fixed this PR according to your comments and resolve a conflict.

File: examples/pruning/tfkeras_integration.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+
+def main():
+
+    study = optuna.create_study(
+        direction='maximize',
+        pruner=optuna.pruners.MedianPruner(n_startup_trials=2)
+    )
+
+    study.optimize(objective, n_trials=10)
```
Comment (sile) [inline examples/pruning/tfkeras_integration.py]: I think that `n_trials=10` is too few (the default `TPESampler` samples parameters at random when the number of finished trials is less than 10).
How about changing the value of `n_trials` to `25`?
Besides, recently, I prefer to specify `timeout` argument to ensure that an example finishes within a reasonable time even if it is executed on a poor CPU/GPU machine.
```suggestion
    study.optimize(objective, n_trials=25, timeout=600
```

**Comment by sfujiwara** [comment]
@sile Thanks! I fixed `n_trials`.

**Pull request merged**